### ISSUE TITLE
Unity ASA 2.9.0: Inconsistent Anchor Location on UWP

### ISSUE THREAD

**Issue by Connor-Kerwin** [Azure/azure-spatial-anchors-samples#252]
**Description**

Anchors are sometimes not locating correctly, with an invalid AR Anchor being generated. I believe it's related to neighbourhood anchors specifically. 

I'm listening to OnAnchorLocated in the SpatialAnchorManager and using this event to attach an object to an anchor. Through instrumenting the code, I've noticed that when it locates the first anchor it seems to be passed with the "Located" status and works correctly, however the other anchors are sometimes being given "Already Tracked" status, which results in them not appearing correctly.

**Development information (please complete the following information)**

 - OS: Windows
 - Development Platform\Version: Visual Studio 2019, Unity 2020.3.1f1
 - Azure Spatial Anchors SDK version: 2.9.0
 - Build Configuration: ARM64 Release

**AR Device information (please complete the following information):**

 - Device: HoloLens 2
 - OS: ARM

**Comment by phamqduc** [comment]
Thank you for letting us know the concerns. We have a few questions below to help us understand the issues.

1) Were the anchors with "Already Tracked" status previously **located** in the same session?
2) Were the anchors with "Already Tracked" status previously **created** in the same session?
3) Did the anchors with "Already Tracked" status **not** appear in correct locations?
4) If so, were they previously reported in correct positions?
5) During the period from the last "Located" status with correct position until the next "Already Tracked" status with incorrect appearances, were there any tracking drifts?
6) Are you able to observe similar issue in the sample app(s)?
If so, could you please provide details of the sequence of moving HLs, creating/retrieving the anchors as well as any code change?
If you can also provide us with the SDK session IDs and anchor IDs, we can also inspect the log to see any issue.

I'm looking forward to hearing from you.

**Comment by Connor-Kerwin** [comment]
I've just done a clean test to reproduce the issue for you.

As a brief overview, the application creates anchors with ASA and stores the identifiers externally so that the placed anchors can be queried by identifier when restarted later for relocation. I ran it three times, first to create the anchors, then two other reruns for locating and demonstrating the inconsistent location behaviour; during the reruns, no other anchors were created. 

The results can be seen below:

<table>
<tr>
<th width="30%">Session Identifier</th>
<th width="50%">Anchors</th>
<th>Results</th>
</tr>
<tr>
<td>155ff3ee-a1b6-43c1-90aa-6a6c8371af1a</td>
<td>

```
a65e2912-cfc1-451e-ae1a-8d156de62e3a - Created and located

7cfc472e-25d4-49ee-8c91-0104982cebba - Created and located

c623ce43-c0d5-4de3-9d23-8bf0ebfeebe4 - Created and located

a2d4e48c-b14b-4ae3-bf95-ea875f262da7 - Created and located

03615d41-0c81-4dac-bb9d-09e42ba345a9 - Created and located
```
</td>
<td>
All 5 anchors created correctly (no locate ops. because the anchors were placed on the device)
</td>
</tr>
<tr>
<td>ce61948d-9079-49b4-83eb-7ff805540948</td>
<td>
  
```
a65e2912-cfc1-451e-ae1a-8d156de62e3a - Located

7cfc472e-25d4-49ee-8c91-0104982cebba - Located

c623ce43-c0d5-4de3-9d23-8bf0ebfeebe4 - Located

a2d4e48c-b14b-4ae3-bf95-ea875f262da7 - Already tracked

03615d41-0c81-4dac-bb9d-09e42ba345a9 - Already tracked
```
</td>
<td>
Created the 3 'located' anchors correctly.
Failed to generate anything for the 2 'already tracked' anchors
</td>
</tr>
<tr>
<td>8061bbfc-1cc2-46ae-b1c2-5f84b0f67008</td>
<td>
  
```
a65e2912-cfc1-451e-ae1a-8d156de62e3a - Located

7cfc472e-25d4-49ee-8c91-0104982cebba - Located

c623ce43-c0d5-4de3-9d23-8bf0ebfeebe4 - Already tracked

a2d4e48c-b14b-4ae3-bf95-ea875f262da7 - Already tracked

03615d41-0c81-4dac-bb9d-09e42ba345a9 - Already tracked
```
</td>
<td>
Created the 2 'located' anchors correctly.
Failed to generate anything for the 3 'already tracked' anchors
</td>
</tr>

</table>

I have the log files for these tests available, however they contain sensitive information so I'd prefer to send those directly to you if needed.

I've not managed to use the sample applications for comparison, so I'm unfortunately not able to provide that information currently.

While investigating the issue, I did also notice that the ASA SpatialAnchorManager OnAnchorLocated method only creates Unity ARAnchors if the event status is 'Located', this may potentially be the cause, as I'm not getting any ARAnchors generate from the 'Already Tracked' anchors from the tests.

I hope this helps, let me know if you have any other questions.

**Comment by phamqduc** [comment]
Thank you so much for the clean test and its details. We will look further into these sessions and anchors and get back to you.

**Comment by phamqduc** [comment]
Thanks again for the information. We have checked on the server logs the anchor's positions between the last two runs that located the anchors and their positions appear quite consistent. In particular, the relative distances (and angles) between the anchors are staying almost the same between the locate runs.
Moreover, positions for the latter four anchors are all determined based on their connections to the first anchor and returned to the application. It's not clear to me yet why they may behave differently.

Do you observe that some anchors (e.g. 03615d41-0c81-4dac-bb9d-09e42ba345a9) are never returned with the status "Located" ? 
Please also let us know if you are able to re-produce the issue in the sample app(s).

**Comment by Connor-Kerwin** [comment]
I did a few more tests to give you some more data to work from in regards to your question about the other anchors. It seems like a client-side issue and potentially a race condition? One of the tests I did located all anchors correctly.

Going back to what I said before with the OnAnchorLocated event in the Unity SpatialAnchorManager, in all cases where 'already tracked' is passed through, it doesn't create the Unity AR Anchor, which would explain why I'm not seeing these anchors appear in app. It sounds like an event maybe isn't being fired from the native plugin to notify creation of an anchor locally for neighbour anchors?

I've not yet used the sample apps to reproduce the issue, but I can have a look if it would help.

<table>
<tr>
<th width="30%">Session Identifier</th>
<th width="50%">Anchors</th>
<th>Results</th>
</tr>
<tr>
<td>8f3f8cf5-4405-454b-b3b7-3c8d5e552ade</td>
<td>

```
a65e2912-cfc1-451e-ae1a-8d156de62e3a - Located

7cfc472e-25d4-49ee-8c91-0104982cebba - Located

c623ce43-c0d5-4de3-9d23-8bf0ebfeebe4 - Already tracked

a2d4e48c-b14b-4ae3-bf95-ea875f262da7 - Already tracked

03615d41-0c81-4dac-bb9d-09e42ba345a9 - Already tracked
```
</td>
<td>
Created the 2 located anchors correctly.
Failed to generate anything for the 3 already tracked anchors.
</td>
</tr>
<tr>
<td>f4ae50d9-07be-487f-be01-a95ffc7618e8</td>
<td>
  
```
a65e2912-cfc1-451e-ae1a-8d156de62e3a - Located

7cfc472e-25d4-49ee-8c91-0104982cebba - Located

c623ce43-c0d5-4de3-9d23-8bf0ebfeebe4 - Located

a2d4e48c-b14b-4ae3-bf95-ea875f262da7 - Located

03615d41-0c81-4dac-bb9d-09e42ba345a9 - Located
```
</td>
<td>
Created and located all anchors correctly.
</td>
</tr>
<tr>
<td>0d803510-72d7-4b43-9bb1-eb501b1f3a0c</td>
<td>
  
```
a65e2912-cfc1-451e-ae1a-8d156de62e3a - Already tracked

7cfc472e-25d4-49ee-8c91-0104982cebba - Located

c623ce43-c0d5-4de3-9d23-8bf0ebfeebe4 - Located

a2d4e48c-b14b-4ae3-bf95-ea875f262da7 - Located

03615d41-0c81-4dac-bb9d-09e42ba345a9 - Already tracked
```
</td>
<td>
Created the 3 located anchors correctly.
Failed to generate anything for the 2 already tracked anchors.
</td>
</tr>
<tr>
<td>ed1a59cc-73bb-4b45-9973-07c2cd5b1eeb</td>
<td>
  
```
a65e2912-cfc1-451e-ae1a-8d156de62e3a - Already tracked

7cfc472e-25d4-49ee-8c91-0104982cebba - Already tracked

c623ce43-c0d5-4de3-9d23-8bf0ebfeebe4 - Already tracked

a2d4e48c-b14b-4ae3-bf95-ea875f262da7 - Located

03615d41-0c81-4dac-bb9d-09e42ba345a9 - Already tracked
```
</td>
<td>
Created 1 anchor correctly.
Failed to generate anything for the 4 already tracked anchors.
</td>
</tr>

</table>

**Comment by Connor-Kerwin** [comment]
I've managed to make all the anchors appear consistently by editing the SpatialAnchorManager and making it check for both Located and AlreadyTracked statuses within OnAnchorLocated when it's creating the AR Anchors. This would obviously break quite easily if an anchor is located again, resulting in multiple AR Anchors, but hopefully this helps towards tracking down the root cause of the issue?

**Comment by phamqduc** [comment]
Thank you so much! We have discussed the case further in our team and have a few more questions regarding the troublesome behaviors that you observed:

- How do you implement the restart/reset for the session in-between "runs"? 
- Could you also try to to set "bypasscache" to "true" to see if the issue still occurs? An example of setting is [here](https://github.com/Azure/azure-spatial-anchors-samples/blob/08e00c61d419e2e322be883f9adf3ee8bf0eb0ed/Unity/Assets/AzureSpatialAnchors.Examples/Scripts/DemoScriptBase.cs#L262). Description of the field is [here](https://docs.microsoft.com/en-us/azure/spatial-anchors/concepts/anchor-locate-strategy#bypasscache). 
- Did you also observe similar behavior in the app on previous versions of the ASA SDK ? Is it a behavior change in SDK 2.9.0 ?

There are a couple of more information we would like to seek
1. version numbers for the ARFoundation and WindowsMR plugins
2. diagnostics logs - please enable diagnostics logging and submit them according to [this page](https://docs.microsoft.com/en-us/azure/spatial-anchors/concepts/logging-diagnostics#enable-diagnostics-logging).

We really appreciate your responses that help us troubleshoot the issue.

**Comment by Connor-Kerwin** [comment]
To answer some of your further questions:

* A "run" is a clean execution of the application, then I restart it entirely, so there is no restart/reset behaviour in-app.
* I've just configured the app to use bypass cache and it does seem to have a difference. Through 4 runs, I didn't have any invalid anchors.
* I never noticed similar behaviour pre 2.9.0. I did expect some minor issues with this release though, due to the change from Legacy XR World Anchors to AR Foundation AR Anchors.

As for extra information:
* ARFoundation: 4.0.12
* Windows XR Plugin: 4.4.1

For diagnostic logs, do you require logs with bypass caching on or off?

**Comment by Connor-Kerwin** [comment]
I've submitted diagnostics for three sessions. Each session had different anchors being located, which should provide you some useful data, all of which failed to locate all anchors. The session Id's are listed below:

* a3af6879-eac3-47b6-a0db-375ddb9f59de
* 4a840581-48bf-44e8-9c7c-290f6a379e92
* 64dc18b6-7288-4803-8078-668501c07d68

I didn't enable image capture within the diagnostics, but if this is required, please let me know and I'll generate some more. Hopefully those were uploaded successfully for you to access.

**Comment by msftradford** [comment]
Thanks for all of the information, we're working on a fix for this issue (#36539781) which should be included in the next release.

**Comment by phamqduc** [comment]
ASA SDK 2.11.0 was released recently that addresses the bug #36539781 mentioned above. Please give SDK 2.11 a try and let us know if you still experience the issue. Thanks.
